feat: header bidding

miguelpeixe opened the pull request:
# Experimental header bidding support

This PR implements an experimental client-side header bidding support using [Prebid.js](https://prebid.org/) with the required modules for connecting with [Media.net](https://www.media.net/) as default bidder.

![image](https://user-images.githubusercontent.com/820752/144904676-4ca75bec-df4e-4984-8f9c-146187a28491.png)

## Implementation details

### Settings

Header bidding configuration settings are implemented using the `Newspack_Ads_Settings` feature. Additional bidder/partner configuration is also added using this system.

### Bidders/Partners

The bidder, or partner, is registered through the `newspack_register_ads_bidder()` function. [The default partner is implemented separately](https://github.com/Automattic/newspack-ads/pull/231/files#diff-c2ab2e6206485804b55dbfd303d2b072a7bd0f51cfa5747793332fbc78fea58f) for maintainability and also to provide an example of registering new bidders natively or through other plugins.

After registering through `newspack_register_ads_bidder()`, it should also hook into the [`newspack_ads_{$bidder_id}_ad_unit_bid`](https://github.com/Automattic/newspack-ads/pull/231/files#diff-46ecf0155737008dc15ef7553e83f14a15b501eb8b38f8d3cb326543947de987R185-R197) filter to apply the appropriate Prebid.js ad unit configuration. The default bidder also [provides a good example](https://github.com/Automattic/newspack-ads/pull/231/files#diff-c2ab2e6206485804b55dbfd303d2b072a7bd0f51cfa5747793332fbc78fea58fR74-R95) of how it works.

### Placements

The placements system has been extended to implement the bidder "placement ID". It's a unique ID provided by the bidder that should be used for identifying the exact bid with the provider.

Currently, each registered bidder generates an automatic `{$bidder_name} Placement ID` field inside a placement. It does not support extra bidder parameters per placement, as it has not yet been identified a reason to support such customization.

While selecting ad units, the UI detects if the registered bidder supports the selected ad unit sizes. If no size is supported, the bidder placement ID gets disabled and a warning is shown. It shouldn't affect the regular behavior of the ad unit and GPT, but also shouldn't attempt bidding.

![image](https://user-images.githubusercontent.com/820752/144490826-98ef9858-8962-4c15-b8d2-128ecf414f4c.png)

### GPT script

Hooks have been added to the GPT ad units and script configuration so that the header bidding implementation can append ad unit metadata and inject the Prebid.js script.

### Wizard UI

The placements UI has been updated to accommodate the increase of fields. The modal approach has been chosen to not flood the wizard screen with too many form fields and buttons.

![image](https://user-images.githubusercontent.com/820752/144487236-9a265a6a-b1c5-48c5-bc24-9c0dc3d6a1d1.png)

![image](https://user-images.githubusercontent.com/820752/144487277-a7c61982-6924-421e-a62a-0fe92b368764.png)

![image](https://user-images.githubusercontent.com/820752/144487359-6d551ce2-19b3-4cb5-9cff-662aadafed10.png)

## Experimental

This is an experimental feature but should still be able to function while disabled. For that reason, the method we'll temporarily use for disabling it is to consider that there are no registered bidders available.

To enable the bidder registration you must add this to your `wp-config.php`:

```php
define( 'NEWSPACK_ADS_EXPERIMENTAL_MEDIANET', true );
```

## Testing

The first test is to make sure the changes here don't affect ads configuration or display, other than the placements UI updates.

**Before adding the `NEWSPACK_ADS_EXPERIMENTAL_MEDIANET` constant**:

1. Check out this branch and https://github.com/Automattic/newspack-plugin/pull/1285
2. Visit **Advertising -> Settings**
3. Ensure no Header Bidding section is visible
4. Visit **Advertising -> Placements**
5. Confirm you see the new UI as shown on the images above
6. Click on the "Placement settings" button of a placement, confirm no "Media.net" field appears
7. Select an ad unit and save
8. Confirm that the placement works as expected with the selected ad unit

### Header bidding settings

1. Define `NEWSPACK_ADS_EXPERIMENTAL_MEDIANET` as described above
2. Visit **Advertising -> Settings**
3. Enable **Header Bidding** and do not change any setting just yet
4. Visit **Advertising -> Placements**, edit a placement, and confirm there's still **no Media.net field**. It should still require the **Media.net Customer ID**.
5. Back to **Advertising -> Settings**, fill the **Media.net Customer ID** with a random `1234567890`, enable **Debug mode**, and click **Save settings**
6. Visit **Advertising -> Placements**, edit a valid placement and confirm you see the **Media.net Placement ID** field


### Placement editing and debug mode

1. Still editing the placement, select different ad units
2. Confirm you see the warning message when ad units with non-standard sizes are selected
3. Save the placement with a valid ad unit and a random `1234567890` as the **Media.net Placement ID**
4. Visit a page where this placement renders the ad, open `DevTools` and confirm there are no errors and you see debug logs similar to this:

![image](https://user-images.githubusercontent.com/820752/144493713-a42b0970-7181-4fd8-be2e-c36b970192ce.png)

No actual bids should take place, as we are using fake IDs.

### Unit tests

[Written unit tests](https://github.com/Automattic/newspack-ads/pull/231/files#diff-f69462aa64c5b2a82c9734232ea41161a6a43231ac6ec38ac9c67a0a877a931e) should also pass.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201230241510067

kmwilkerson commented:
This all looks wonderful so far @miguelpeixe!! 

> ![image](https://user-images.githubusercontent.com/20666833/144681566-575bf8c9-2563-456c-9d97-2007fcd1fa47.png)
- Can we set our default and recommended price granularity to "dense" (it has the best results and there's less wastage).
- Update price granularity explanation to:
"Price granularity is the level of the price target set in the header bidding line items. The price buckets impact the line items’ ability to compete optimally with your other price-based ad campaigns."

**Pushing Order/Line Items to GAM**
As we talked about in Slack, we need to push the required line items to GAM for header bidding, and therefore will require OAuth for anyone wanting to enable header bidding. We have two options: one set of line items for all bidders OR a set of line items for each bidder. The first option would be the simplest/less error-prone and the trigger for sending the line items could be clicking the save button after they select the price granularity. This would be the only time we'd need to push line items, as all bidders would feed out of the same order and no future changes to those line items would be needed. Noting that this route would require us to set key `bidder=bidder_name` as a custom dimension. It's a very small detail, but is important not to forget.

miguelpeixe commented:
> * Can we set our default and recommended price granularity to "dense" (it has the best results and there's less wastage).
> * Update price granularity explanation to:
>   "Price granularity is the level of the price target set in the header bidding line items. The price buckets impact the line items’ ability to compete optimally with your other price-based ad campaigns."

Thanks @kmwilkerson ! 5d7a3623471aa9842ce8386f81f7903782774d51 updates the default value and the help text for "Price granularity". I hope it's ok I removed the initial "Price granularity is" since it has already been stated by the field label. I've updated the screenshot above to represent the change.

> will require OAuth for anyone wanting to enable header bidding.

Does the manual, or custom script, approach for creating the line items would not work for users without OAuth? I agree that we should have this implemented as part of the setup process when available.

I'd rather work on this in a different PR since it's not a dependency for our experimental testings and should require a dedicated review on the implementation.

> We have two options: one set of line items for all bidders OR a set of line items for each bidder.

Can we have a separate issue to continue this discussion?

kmwilkerson commented:
I'll open up a new issue to continue GAM discussion! Thanks @miguelpeixe!

dkoo commented on tests/test-settings.php:
```diff
@@ -141,6 +143,18 @@ public function test_update_value_within_options() {
+		self::assertSame(
+			$settings['select_field'],
+			'option2',
+			'Should update a value withing existing options.'
```
```suggestion
			'Should update a value within existing options.'
```

dkoo commented on includes/class-newspack-ads-bidding.php:
```diff
@@ -0,0 +1,558 @@
+			if ( isset( $ad_data['bidders'] ) && count( $ad_data['bidders'] ) ) {
+
+				// Detect sizes supported by available bidders.
+				$sizes = array_intersect( $ad_data['sizes'], $bidders_sizes );
```
Also ran into another PHP error triggered by this line on both my local and the new site:

```bash
[08-Dec-2021 19:15:19 UTC] PHP Notice:  Array to string conversion in /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-bidding.php on line 172
[08-Dec-2021 19:15:19 UTC] PHP Stack trace:
[08-Dec-2021 19:15:19 UTC] PHP   1. {main}() /Users/dkoo/Local Sites/newspack/app/public/index.php:0
[08-Dec-2021 19:15:19 UTC] PHP   2. require() /Users/dkoo/Local Sites/newspack/app/public/index.php:17
[08-Dec-2021 19:15:19 UTC] PHP   3. require_once() /Users/dkoo/Local Sites/newspack/app/public/wp-blog-header.php:19
[08-Dec-2021 19:15:19 UTC] PHP   4. include() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template-loader.php:106
[08-Dec-2021 19:15:19 UTC] PHP   5. get_footer() /Users/dkoo/Local Sites/newspack/app/public/wp-content/themes/repo/newspack-theme/front-page.php:41
[08-Dec-2021 19:15:19 UTC] PHP   6. locate_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/general-template.php:92
[08-Dec-2021 19:15:19 UTC] PHP   7. load_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:716
[08-Dec-2021 19:15:19 UTC] PHP   8. require_once() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:770
[08-Dec-2021 19:15:19 UTC] PHP   9. wp_footer() /Users/dkoo/Local Sites/newspack/app/public/wp-content/themes/repo/newspack-theme/footer.php:67
[08-Dec-2021 19:15:19 UTC] PHP  10. do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/general-template.php:3031
[08-Dec-2021 19:15:19 UTC] PHP  11. WP_Hook->do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/plugin.php:470
[08-Dec-2021 19:15:19 UTC] PHP  12. WP_Hook->apply_filters() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:327
[08-Dec-2021 19:15:19 UTC] PHP  13. Newspack_Ads_Blocks::insert_google_ad_manager_footer_code() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:303
[08-Dec-2021 19:15:19 UTC] PHP  14. do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-blocks.php:206
[08-Dec-2021 19:15:19 UTC] PHP  15. WP_Hook->do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/plugin.php:470
[08-Dec-2021 19:15:19 UTC] PHP  16. WP_Hook->apply_filters() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:327
[08-Dec-2021 19:15:19 UTC] PHP  17. Newspack_Ads_Bidding->prebid_script() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:303
[08-Dec-2021 19:15:19 UTC] PHP  18. array_intersect() /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-bidding.php:172
```

miguelpeixe replied:
> The warning about unsupported sizes only seems to appear if all of the sizes in an ad unit are unsupported—is this intended?

It is. If there's any size available for bidding, it should use it.

There was a logic error on how the `$bidders_sizes` was formed, it should be fixed by 6229c3b6a1d406988ff4ff9136dd051a09e27e0b

dkoo replied:
This is still throwing the warning for me. I think the issue is that we're using `array_intersect` to compare two arrays which consist of sub-arrays of sizes. According to the PHP docs for `array_interesect`:

> Note: Two elements are considered equal if and only if (string) $elem1 === (string) $elem2. In words: when the string representation is the same. 

Assuming the `$ad_data['sizes']` and `$bidder_sizes` arrays always look something like this, with this exact level of nesting:

```php
Array
(
    [0] => Array
        (
            [0] => 320
            [1] => 50
        )

    [1] => Array
        (
            [0] => 320
            [1] => 100
        )

    [2] => Array
        (
            [0] => 728
            [1] => 90
        )

    [3] => Array
        (
            [0] => 970
            [1] => 90
        )
)
```

We would need to stringify the sub-arrays to compare the parent arrays, then de-stringify on the final output. Something like this:

```php
// Detect sizes supported by available bidders.
$sizes = array_map(
	function ( $matched_size ) {
		return json_decode( $matched_size );
	},
	array_intersect(
		array_map(
			function( $size ) {
				return json_encode( $size );
			},
			$ad_data['sizes']
		),
		array_map(
			function( $size ) {
				return json_encode( $size );
			},
			$bidders_sizes
		),
	)
);
```

We may need additional sorting to ensure that the stringified size values can be compared properly.

miguelpeixe replied:
Good catch! 7b433684e4600713b0853030f1d1b578719899ae implements some mapping for the custom typecasting

dkoo replied:
That did it—thanks for the fix!

dkoo commented on includes/class-newspack-ads-placements.php:
```diff
@@ -311,27 +410,46 @@ public static function disable_placement( $placement_key ) {
+
 		if ( $hook_key ) {
-			$ad_unit_key = 'ad_unit_' . $hook_key;
+			$placement_data = $placement['data']['hooks'][ $hook_key ];
```
Ran into a PHP error triggered by this line on both my local site and a fresh site. It seems to happen when loading up placements on the front-end which lack a `hooks` key, such as `global_above_header` which has a `hook_name` key but not a `hooks` key.

```bash
PHP Notice:  Undefined index: hooks in /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-placements.php on line 433
[08-Dec-2021 19:15:18 UTC] PHP Stack trace:
[08-Dec-2021 19:15:18 UTC] PHP   1. {main}() /Users/dkoo/Local Sites/newspack/app/public/index.php:0
[08-Dec-2021 19:15:18 UTC] PHP   2. require() /Users/dkoo/Local Sites/newspack/app/public/index.php:17
[08-Dec-2021 19:15:18 UTC] PHP   3. require_once() /Users/dkoo/Local Sites/newspack/app/public/wp-blog-header.php:19
[08-Dec-2021 19:15:18 UTC] PHP   4. include() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template-loader.php:106
[08-Dec-2021 19:15:18 UTC] PHP   5. get_header() /Users/dkoo/Local Sites/newspack/app/public/wp-content/themes/repo/newspack-theme/front-page.php:10
[08-Dec-2021 19:15:18 UTC] PHP   6. locate_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/general-template.php:48
[08-Dec-2021 19:15:18 UTC] PHP   7. load_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:716
[08-Dec-2021 19:15:18 UTC] PHP   8. require_once() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:770
[08-Dec-2021 19:15:18 UTC] PHP   9. get_template_part() /Users/dkoo/Local Sites/newspack/app/public/wp-content/themes/repo/newspack-theme/header.php:39
[08-Dec-2021 19:15:18 UTC] PHP  10. locate_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/general-template.php:204
[08-Dec-2021 19:15:18 UTC] PHP  11. load_template() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:716
[08-Dec-2021 19:15:18 UTC] PHP  12. require() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/template.php:772
[08-Dec-2021 19:15:18 UTC] PHP  13. dynamic_sidebar() /Users/dkoo/Local Sites/newspack/app/public/wp-content/themes/repo/newspack-theme/template-parts/header/mobile-sidebar.php:34
[08-Dec-2021 19:15:18 UTC] PHP  14. do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/widgets.php:724
[08-Dec-2021 19:15:18 UTC] PHP  15. WP_Hook->do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/plugin.php:470
[08-Dec-2021 19:15:18 UTC] PHP  16. WP_Hook->apply_filters() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:327
[08-Dec-2021 19:15:18 UTC] PHP  17. Newspack_Ads_Sidebar_Placements::create_sidebar_before_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:303
[08-Dec-2021 19:15:18 UTC] PHP  18. do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-sidebar-placements.php:46
[08-Dec-2021 19:15:18 UTC] PHP  19. WP_Hook->do_action() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/plugin.php:470
[08-Dec-2021 19:15:18 UTC] PHP  20. WP_Hook->apply_filters() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:327
[08-Dec-2021 19:15:18 UTC] PHP  21. Newspack_Ads_Placements::{closure:/Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-placements.php:183-185}() /Users/dkoo/Local Sites/newspack/app/public/wp-includes/class-wp-hook.php:303
[08-Dec-2021 19:15:18 UTC] PHP  22. Newspack_Ads_Placements::inject_placement_ad() /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-ads/includes/class-newspack-ads-placements.php:184
```

It's not happening at this moment on the fresh site, so I'm not sure what changed there, but I do see the error in the debug.log, so it was happening at some point during testing.

miguelpeixe replied:
It was missing another check to ensure that there is placement data for the hook. Fixed in 02b3606d4465841448a9a7e24e3bf225e0cdf122

dkoo replied:
Confirmed, thanks!

dkoo reviewed (changes requested):
Note: this feature also requires AMP Plus, which is currently enabled using the following environment variable:

```php
define( 'NEWSPACK_AMP_PLUS_ENABLED', true );
```

Ran into a JS error on this step:

> 4. Visit a page where this placement renders the ad, open DevTools and confirm there are no errors and you see debug logs similar to this

```
TypeError: map.entries is not a function
    updateTargetingFromMap https://securepubads.g.doubleclick.net/tag/js/gpt.js:211
    ...
```

Confirmed this on both my local test site and a fresh test site.

EDIT: It appears this JS error happens only on Firefox when using that browser's [enhanced tracking protection](https://support.mozilla.org/en-US/kb/enhanced-tracking-protection-firefox-desktop), which is on by default. This indicates that Firefox may block header bidding by default...will be interesting to test.

Other comments below.

miguelpeixe commented on includes/class-newspack-ads-bidding.php:
```diff
@@ -0,0 +1,558 @@
+	const PREBID_SCRIPT_HANDLE = 'newspack-ads-prebid';
+
+	// Standard sizes accepted by partners.
+	const ACCEPTED_AD_SIZES = [
```
Good call! I filed #256 so we can have another PR synced with the required changes on `newspack-plugin`.

miguelpeixe commented on includes/bidders/class-newspack-ads-bidder-medianet.php:
```diff
@@ -0,0 +1,97 @@
+
+		newspack_register_ads_bidder(
+			'medianet',
+			array(
```
Done in 8f8dfd58c8411d54afcec7fd8e02d8aeec174957

miguelpeixe commented on includes/class-newspack-ads-bidding.php:
```diff
@@ -0,0 +1,558 @@
+						bidsBackHandler: initAdserver,
+					} );
+					/**
+					 * GAM Express Module
```
Done in e92fc148081d76453943c140509f290f35017686

dkoo reviewed (changes requested):
Just one persistent PHP warning. Not sure if it actually prevents the output of `array_intersect` but just in case!

dkoo reviewed (approved):
Looking good now 👍

miguelpeixe commented:
Awesome! Thank you for reviewing this, @dkoo! 🙇 🎉
